java/src/jmri/jmrit/logix/NXFrame.java

```diff
@@ -828,91 +822,99 @@ private String makeCommands(Warrant w) {
-
-            // run through mid route at max speed
-            while (nextIdx < orders.size() && remTotal - blockLen > dnRampLength) {
-                remTotal -= blockLen;
-                // constant speed, get time to next block
-                noopTime = _speedUtil.getTimeForDistance(curThrottle, blockLen);   // time to next block
-                curDistance += blockLen;
-                if (log.isDebugEnabled()) {
-                    log.debug("Leave MidRoute block \"{}\" noopTime= {} blockLen= {} curDistance={} remTotal= {}",
-                            blockName, noopTime, blockLen, curDistance, remTotal);
-                }
-                bo = orders.get(nextIdx++);
-                blockLen = getPathLength(bo);
-                if (blockLen <= 0) {
-                    return Bundle.getMessage("zeroPathLength", bo.getPathName(), bo.getBlock().getDisplayName());
-                 }
-                blockName = bo.getBlock().getDisplayName();
-                if (nextIdx == orders.size()) {
-                    blockLen = _stopDist;
-                }
-                w.addThrottleCommand(new ThrottleSetting(noopTime, Command.NOOP, -1, ValueType.VAL_NOOP, 
-                        SpeedStepMode.UNKNOWN, 0, blockName, _speedUtil.getTrackSpeed(curThrottle)));
-                if (log.isDebugEnabled()) {
-                    log.debug("cmd#{}. Enter MidRoute block \"{}\" noopTime= {}, blockLen= {}, curDistance={}",
-                            cmdNum++, blockName, noopTime, blockLen, curDistance);
+        if (totalLen - sumBlkLen - blockLen > dnRampLength) {   // (sumBlkLen + blockLen) is total distance traveled to end of current block
+            if (!iter.hasNext()) {  // upRamp done. At maxThrottle for remainder of block
+                if (nextIdx < orders.size()) {    // not the last block
+                    if (blkDistance >= blockLen) {
+                        curDistance += blockLen; // noop distance
+                    } else {
+                        curDistance += blockLen - blkDistance; // noop distance
+                    }
+                    sumBlkLen += blockLen;
+                    bo = orders.get(nextIdx++);
+                    blockLen = getPathLength(bo);
+                    if (blockLen <= 0) {
+                        return Bundle.getMessage("zeroPathLength", bo.getPathName(), bo.getBlock().getDisplayName());
+                     }
+                    blockName = bo.getBlock().getDisplayName();
+                    w.addThrottleCommand(new ThrottleSetting(noopTime, Command.NOOP, -1, ValueType.VAL_NOOP, 
+                            SpeedStepMode.UNKNOWN, 0, blockName, _speedUtil.getTrackSpeed(curThrottle)));
+                    if (log.isDebugEnabled()) {
+                        log.debug("cmd#{}. Ramp Up done at block \"{}\" noopTime= {}, blockLen= {}, sumBlkLen= {}, curDist= {}",
+                                cmdNum++, blockName, noopTime, speedTime, blockLen, sumBlkLen, curDistance);
```

Comment:
The log statement has 6 placeholders {} but 7 parameters. The parameter `speedTime` doesn't have a placeholder.